### PR TITLE
Implements Protocore::proposeLinkInternal

### LINKED ISSUE #181: Implement Protocore::_proposeLink()
1. Implement `Protocore::_proposeLink()` 
    - This will be an internal function
    - This function proposes a valid link to be voted later by active validators.
    - Checks if the `parentVoteMessageHash` is 0.
    - Checks if the `parentVoteMessageHash` refers to an already proposed link which `targetFinalisation` is at least justified.
    - Checks if `targetBlockHash` is 0
    - Checks if `targetBlockNumber` is a multiple of the epoch length.
    - Checks if `targetBlockNumber` is bigger than a targetBlockNumber pointed by `_parentVoteMessageHash` link.
    - Proposal
    ```solidity
    function _proposeLink(
        bytes32 _parentVoteMessageHash,
        bytes32 _sourceTransitionHash,
        bytes32 _targetBlockHash,
        uint256 _targetBlockNumber
    )
        internal
    ```

### PR DESCRIPTION
**Pull request by pgev**
Fixes #181

### CODE REVIEW

File: contracts/protocore/Protocore.sol [hunk 1]
```diff
@@ -114,4 +141,102 @@ contract Protocore is MosaicVersion, ValidatorSet, CoconsensusModule {
...
+    {
+        require(
+            _parentVoteMessageHash != bytes32(0),
+            "Parent vote message hash is 0."
+        );
+
+        Link storage parentLink = links[_parentVoteMessageHash];
+
+        require(
+            parentLink.targetBlockHash != bytes32(0),
```
Comment (benjaminbollen) [inline contracts/protocore/Protocore.sol]: better to check the status? `parentLink.targetFinalisation >= CheckpointFinalisationStatus.Registered`, then it collapses into the next requirement and we save one

File: contracts/protocore/Protocore.sol [hunk 2]
```diff
@@ -114,4 +141,102 @@ contract Protocore is MosaicVersion, ValidatorSet, CoconsensusModule {
...
+            "Parent link does not exist."
+        );
+
+        require(
+            parentLink.targetFinalisation >= CheckpointFinalisationStatus.Justified,
+            "Parent link's target finalisation status should be at least justified."
+        );
+
+        require(
```
Comment (benjaminbollen) [inline contracts/protocore/Protocore.sol:179]: moving these checks up would save gas on reading from storage; but not so important :D

File: contracts/protocore/Protocore.sol [hunk 3]
```diff
@@ -114,4 +141,102 @@ contract Protocore is MosaicVersion, ValidatorSet, CoconsensusModule {
...
+            parentLink.targetBlockNumber,
+            _targetBlockNumber
+        );
+
+        require(
+            links[voteMessageHash].targetBlockHash == bytes32(0),
+            "The proposed link already exists."
+        );
+
+        Link storage proposedLink = links[voteMessageHash];
```
Comment (benjaminbollen) [inline contracts/protocore/Protocore.sol]: move this above L214 to save gas?

File: contracts/protocore/Protocore.sol [hunk 4]
```diff
@@ -114,4 +141,102 @@ contract Protocore is MosaicVersion, ValidatorSet, CoconsensusModule {
...
+        proposedLink.sourceTransitionHash = _sourceTransitionHash;
+        proposedLink.proposedMetablockHeight = openKernelHeight;
+        proposedLink.targetFinalisation = CheckpointFinalisationStatus.Registered;
+    }
+
+    /**
+     * @notice Takes vote message parameters and returns the typed vote
+     *         message hash.
+     */
+    function hashVoteMessageInternal(
```
Comment (benjaminbollen) [inline contracts/protocore/Protocore.sol:227]: huh ? Im confused, is this to accomodate with the other ticket which is taken up in parallel? 🧐

File: contracts/test/protocore/TestProtocore.sol [hunk 5]
```diff
@@ -14,29 +14,207 @@ pragma solidity >=0.5.0 <0.6.0;
...
+            _epochLength
+        );
     }
 
-    function getCoconsensus() public view returns (CoconsensusI) {
+
+    /* External Functions */
+
+    function getCoconsensus()
+		public
```
Comment (benjaminbollen) [inline contracts/test/protocore/TestProtocore.sol:83]: indentation seems off

File: test/protocore/propose_link_internal.js [hunk 6]
```diff
@@ -0,0 +1,259 @@
+// Copyright 2019 OpenST Ltd.
```
Comment (benjaminbollen) [inline test/protocore/propose_link_internal.js]: Happy new year 2020 🥳

File: test/protocore/propose_link_internal.js [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+          voteMessageHash1,
+          sourceTransitionHash2,
+          targetBlockHash2,
+          targetBlockNumber2,
+        ),
+        'Parent link\'s target finalisation status should be at least justified.',
+      );
+    });
+
+    it('should revert if target block number is not a multiply of the epoch length', async () => {
```
Comment (benjaminbollen) [inline test/protocore/propose_link_internal.js]: ".. is not a multiple of .."

**Review by benjaminbollen — COMMENTED** [review]
minor comments / questions

File: contracts/protocore/Protocore.sol (hunk 1 above)
Comment (pgev) [reply contracts/protocore/Protocore.sol]: Correct. Improved!

File: contracts/protocore/Protocore.sol (hunk 2 above)
Comment (pgev) [reply contracts/protocore/Protocore.sol:179]: True. Improved!

File: contracts/protocore/Protocore.sol (hunk 4 above)
Comment (pgev) [reply contracts/protocore/Protocore.sol:227]: True. I've not wanted to block myself out hence introduced the function interface, made it internal and mocked it in TestProtocore.

File: contracts/test/protocore/TestProtocore.sol (hunk 5 above)
Comment (pgev) [reply contracts/test/protocore/TestProtocore.sol:83]: It is correct in my VS code. Github glitch? ...

File: test/protocore/propose_link_internal.js (hunk 6 above)
Comment (pgev) [reply test/protocore/propose_link_internal.js]: Happy belated New Year 2020 🎆 🤗 
Fixed ...

File: test/protocore/propose_link_internal.js (hunk 7 above)
Comment (pgev) [reply test/protocore/propose_link_internal.js]: Thanks. Fixed.

**Review by deepesh-kn — APPROVED** [review]

**Pull request merged**